File: algorithms/python/valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        st = []
        p = {')': '(', ']': '[', '}': '{'}

        for i in s:
            if i in '([{':
                st.append(i)
            elif i in ')]}':
                if not st or st[-1] != p[i]:
                    return False
                st.pop()
        return not st

File: algorithms/c/valid_parentheses.c
bool isValid(char* s) {
    char st[100000];
    int top = -1;

    for (int i = 0; s[i] != '\0'; i++) {
        char c = s[i];

        if (c == '(' || c == '[' || c == '{') {
            st[++top] = c;
        } else {
            if (top == -1) return false;
            char tc = st[top--];

            if((c == ')' && tc != '(') || (c == ']' && tc != '[') || (c == '}' && tc != '{')) {
                return false;
            }
        }
    }

    return top == -1;
}
